integration-cli/docker_cli_build_test.go

```diff
@@ -1924,6 +1926,121 @@ func TestBuildForceRm(t *testing.T) {
 }
 
+// Test that an infinite sleep during a build is killed if the client disconnects.
+// This test is fairly hairy because there are lots of ways to race.
+// Strategy:
+// * Monitor the output of docker events starting from before
+// * Run a 1-year-long sleep from a docker build.
+// * When docker events sees container start, close the "docker build" command
+// * Wait for docker events to emit a dying event.
+func TestBuildCancelationKillsSleep(t *testing.T) {
+	// TODO(jfrazelle): Make this work on Windows.
+	testRequires(t, SameHostDaemon)
+
+	name := "testbuildcancelation"
+	defer deleteImages(name)
+
+	// (Note: one year, will never finish)
+	ctx, err := fakeContext("FROM busybox\nRUN sleep 31536000", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer ctx.Close()
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	finish := make(chan struct{})
+	defer close(finish)
+
+	eventStart := make(chan struct{})
+	eventDie := make(chan struct{})
+
+	// Start one second ago, to avoid rounding problems
+	startEpoch := time.Now().Add(-1 * time.Second)
+
+	// Goroutine responsible for watching start/die events from `docker events`
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// Watch for events since epoch.
+		eventsCmd := exec.Command(dockerBinary, "events",
+			"-since", fmt.Sprint(startEpoch.Unix()))
+		stdout, err := eventsCmd.StdoutPipe()
+		err = eventsCmd.Start()
+		if err != nil {
+			t.Fatalf("failed to start 'docker events': %s", err)
+		}
+
+		go func() {
+			<-finish
+			eventsCmd.Process.Kill()
+		}()
+
+		var started, died bool
+		matchStart := regexp.MustCompile(" \\(from busybox\\:latest\\) start$")
+		matchDie := regexp.MustCompile(" \\(from busybox\\:latest\\) die$")
+
+		//
+		// Read lines of `docker events` looking for container start and stop.
+		//
+		scanner := bufio.NewScanner(stdout)
+		for scanner.Scan() {
+			if ok := matchStart.MatchString(scanner.Text()); ok {
+				if started {
+					t.Fatal("assertion fail: more than one container started")
+				}
+				close(eventStart)
+				started = true
+			}
+			if ok := matchDie.MatchString(scanner.Text()); ok {
+				if died {
+					t.Fatal("assertion fail: more than one container died")
+				}
+				close(eventDie)
+				died = true
+			}
+		}
+
+		err = eventsCmd.Wait()
+		if err != nil && !IsKilled(err) {
+			t.Fatalf("docker events had bad exit status: %s", err)
+		}
+	}()
+
+	buildCmd := exec.Command(dockerBinary, "build", "-t", name, ".")
+	buildCmd.Dir = ctx.Dir
+	buildCmd.Stdout = os.Stdout
+
+	err = buildCmd.Start()
+	if err != nil {
+		t.Fatalf("failed to run build: %s", err)
+	}
+
+	select {
+	case <-time.After(10 * time.Second):
+		t.Fatal("build did not cancel in a timely fashion")
+	case <-eventStart:
+	}
+	// Now we know the container is running
+
+	err = buildCmd.Process.Kill()
+	if err != nil {
+		t.Fatalf("error killing build command: %s", err)
+	}
+
+	<-eventDie
+	// If we don't get here in a timely fashion, it wasn't killed.
```

Comment:
Thanks pwaller!! Really looking forward to getting this in :)

On Saturday, March 21, 2015, Peter Waller <notifications@github.com> wrote:

> In integration-cli/docker_cli_build_test.go
> <https://github.com/docker/docker/pull/9774#discussion_r26890802>:
>
> > +	}
> > +
> > +	select {
> > +	case <-time.After(10 * time.Second):
> > +		t.Fatal("build did not cancel in a timely fashion")
> > +	case <-eventStart:
> > +	}
> > +	// Now we know the container is running
> > +
> > +	err = buildCmd.Process.Kill()
> > +	if err != nil {
> > +		t.Fatalf("error killing build command: %s", err)
> > +	}
> > +
> > +	<-eventDie
> > +	// If we don't get here in a timely fashion, it wasn't killed.
>
> You've persuaded me :)
>
> I apologise for not responding positively on this point initially, and I
> did not think the status quo was an ideal solution, but I wasn't convinced
> anyone was making suggestions from a thoroughly reasoned standpoint (more a
> "he's blocking the build and it's *this sleep*!"). So I think here is the
> right place to do it and we should define a bound on how long Cancellation
> should take. I'm choosing 30 seconds for now, because some poor person
> might want to run the tests in a very slow environment. Usually it should
> seem instant, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9774/files#r26890802>.
>